core/domain/exp_jobs_one_off.py

```diff
@@ -535,3 +536,41 @@ def reduce(key, values):
         # Combine all values from multiple lists into a single list
         # for that error type.
         yield (key, list(set().union(*final_values)))
+
+
+class ImageDataMigrationJob(jobs.BaseMapReduceOneOffJobManager):
+    """ One-off job for migrating the images in the exploration
+        from the GAE to GCS
+    """
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [file_models.FileSnapshotContentModel]
+
+    @staticmethod
+    def map(file_snapshot_content_model):
+        instance_id = file_snapshot_content_model.get_unversioned_instance_id()
+        filetype = instance_id[instance_id.rfind(".") + 1:]
+
+        if filetype in ALLOWED_IMAGE_EXTENSIONS:
+            filename = instance_id[instance_id.rfind("/") + 1:]
+            exploration_id = (instance_id[1:])[:(instance_id[1:]).find("/")]
+            filepath = (instance_id[1:])[(instance_id[1:]).find("/") + 1:]
+            file_model = file_models.FileModel.get_model(
+                exploration_id, filepath, False)
+
+            if file_model and file_model.deleted:
+                # content = file_model.content
+                # fs = fs_domain.AbstractFileSystem(fs_domain.GcsFileSystem(
+                #   exploration_id))
+                # fs.commit(file_model.user_id, '%s/%s' %('images',
+                #   filename, content)
+                file_model.delete("ADMIN", "Delete the file", False)
+                file_metadatmodel = file_models.FileMetadataModel.get_model(
+                    exploration_id, filepath, False)
+                file_metadatmodel.delete(
+                    "ADMIN", "Delete the filemetada", False)
+                yield(exploration_id, filename)
```

Comment:
What's the point of yielding this?

Maybe instead do `yield ('File copied', 1)` and in the reduce step, do an aggregation to give you the total number of files copied and the full list of error messages (see above).

Reply:
I was actually thinking of testing it by loading a demo exploration and then extracting the image filenames from it. If we yield (exploration_id, filename) then we could verify that all the images in the exploration got shifted or not.
But the approach which you said seems better since that would also include the error messages.